core/client/src/light/call_executor.rs

```diff
@@ -218,11 +216,7 @@ pub fn prove_execution<Block, S, E>(
 
 	// execute method + record execution proof
 	let (result, exec_proof) = executor.prove_at_trie_state(&trie_state, &mut changes, method, call_data)?;
-	let total_proof = init_proof.into_iter()
-		.chain(exec_proof.into_iter())
-		.collect::<HashSet<_>>()
-		.into_iter()
-		.collect();
+	let total_proof = merge_storage_proofs(vec![init_proof, exec_proof]);
```

Comment:
vec allocation can be avoided here, especially since merge_storage_proofs takes an iterator as input.

Reply:
How exactly do you recommend creating this iterator? I think something like `iter::once(init_proof).chain(iter::once(exec_proof))` is a bit clunky.

Reply:
This really doesn't seem like a good reason to sacrifice readability to me. `merge_storage_proofs` internally has to create a new `HashSet` and much larger vector of trie nodes anyway.

Reply:
Of course it is just an alloc among lot of others, I think crazy iter expression are rather common in the code base, but not abusing them is probably a good thing.

Regarding the other comment where we grow memory, you also can use a different approach, I know some poeple like this kind of api, I never really use it but it may be good here.

```rust
struct ProofMerger(Hashset<StorageProof>);

impl ProofMerger {
  fn add(self, proof: StorageProof) {
    self.0.extends(proof);
    self
  }
  fn merge(self) -> StorageProof {
    StorageProof { trie_nodes: self.0.into_iter().collect() }
  }

}
```
in this case would look like 

```rust
let total_proof = ProofMerger(init_proof)
  .add(exec_proof)
  .merge();
```

Reply:
Not that it is really only nitpick, should not prevent merging this PR for me.